lib/srv/discovery/fetchers/db/aws_opensearch.go

```diff
@@ -0,0 +1,160 @@
+type openSearchFetcher struct {
+	awsFetcher
+
+	cfg openSearchFetcherConfig
+	log logrus.FieldLogger
+}
+
+// newOpenSearchFetcher returns a new OpenSearch databases fetcher instance.
+func newOpenSearchFetcher(config openSearchFetcherConfig) (common.Fetcher, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &openSearchFetcher{
+		cfg: config,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: "watch:opensearch",
+			"labels":        config.Labels,
+			"region":        config.Region,
+			"role":          config.AssumeRole,
+		}),
+	}, nil
+}
+
+// Get returns OpenSearch databases matching the watcher's selectors.
+func (f *openSearchFetcher) Get(ctx context.Context) (types.ResourcesWithLabels, error) {
+	domains, err := getOpenSearchDomains(ctx, f.cfg.openSearch)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var eligibleDomains []*opensearchservice.DomainStatus
+	for _, domain := range domains {
+		if !services.IsOpenSearchDomainAvailable(domain) {
+			f.log.Debugf("OpenSearch domain %q is unavailable. Skipping.", aws.StringValue(domain.DomainName))
+			continue
+		}
+
+		eligibleDomains = append(eligibleDomains, domain)
+	}
+
+	if len(eligibleDomains) == 0 {
+		return types.ResourcesWithLabels{}, nil
+	}
+
+	var databases types.Databases
+	for _, domain := range eligibleDomains {
+		tags, err := getOpenSearchResourceTags(ctx, f.cfg.openSearch, domain.ARN)
+
+		if err != nil {
+			if trace.IsAccessDenied(err) {
+				f.log.WithError(err).Debug("No permissions to list resource tags")
+			} else {
+				f.log.WithError(err).Infof("Failed to list resource tags for OpenSearch domain %q.", aws.StringValue(domain.DomainName))
+			}
+		}
+
+		dbs, err := services.NewDatabaseFromOpenSearchDomain(domain, tags)
+		if err != nil {
+			f.log.WithError(err).Infof("Could not convert OpenSearch domain %q configuration to database resource.", aws.StringValue(domain.DomainName))
+		} else {
+			databases = append(databases, dbs...)
+		}
+	}
+
+	applyAssumeRoleToDatabases(databases, f.cfg.AssumeRole)
+	return filterDatabasesByLabels(databases, f.cfg.Labels, f.log).AsResources(), nil
+}
+
+// String returns the fetcher's string description.
+func (f *openSearchFetcher) String() string {
+	return fmt.Sprintf("openSearchFetcher(Region=%v, Labels=%v)",
+		f.cfg.Region, f.cfg.Labels)
+}
+
+// getOpenSearchDomains fetches all OpenSearch domains.
+func getOpenSearchDomains(ctx context.Context, client opensearchserviceiface.OpenSearchServiceAPI) ([]*opensearchservice.DomainStatus, error) {
+	names, err := client.ListDomainNamesWithContext(ctx, nil)
+	if err != nil {
+		return nil, trace.Wrap(libcloudaws.ConvertRequestFailureError(err))
+	}
+
+	req := &opensearchservice.DescribeDomainsInput{DomainNames: []*string{}}
+	for _, domain := range names.DomainNames {
+		req.DomainNames = append(req.DomainNames, domain.DomainName)
+	}
+
+	domains, err := client.DescribeDomainsWithContext(ctx, req)
```

Comment:
It is really unfortunate that names have to be listed first then fetch them. Sigh...

Reply:
True, but OTOH, this isn't an interactive workload, so the extra call is not really a performance issue.

Reply:
```
// Container for the parameters to the DescribeDomains operation.
type DescribeDomainsInput struct {
	_ struct{} `type:"structure"`

	// Array of OpenSearch Service domain names that you want information about.
	// If you don't specify any domains, OpenSearch Service returns information
	// about all domains owned by the account.
	//
	// DomainNames is a required field
	DomainNames []*string `type:"list" required:"true"`
}
```

Do we really need the extra call?
> If you don't specify any domains, OpenSearch Service returns information about all domains owned by the account.

Isn't it what we want? all domains

Reply:
The docs appear incorrect in this case. 

If you comment out these lines:
```
	for _, domain := range names.DomainNames {
		req.DomainNames = append(req.DomainNames, domain.DomainName)
	}
```

The request returns an empty domain list, contrary to expectations.

Reply:
Indeed :+1: 
I also tested using the `aws` cli and got the same result: empty list